File: ports/go/src/core/kernel/bcstextstream.go
package kernel

import (
	"bytes"
	"io"
	"strings"
	"sync"
)

type BcsTextStream struct {
	mu     sync.Mutex
	reader io.Reader
	writer io.Writer
	buffer *bytes.Buffer
}

func NewBcsTextStream() *BcsTextStream {
	buf := new(bytes.Buffer)
	return &BcsTextStream{
		reader: buf,
		writer: buf,
		buffer: buf,
	}
}

func (s *BcsTextStream) WriteString(text string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.writer != nil {
		s.writer.Write([]byte(text))
	}
}

func (s *BcsTextStream) ReadAll() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.buffer != nil {
		return s.buffer.String()
	}
	if s.reader != nil {
		b, _ := io.ReadAll(s.reader)
		return string(b)
	}
	return ""
}

func (s *BcsTextStream) ReadLine() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.buffer != nil {
		line, _ := s.buffer.ReadString('\n')
		return strings.TrimSuffix(line, "\n")
	}
	return ""
}

func (s *BcsTextStream) Flush() {
	// Not strictly needed for bytes.Buffer
}

File: ports/rust/src/core/kernel/bcs_textstream.rs
use std::sync::{Arc, Mutex};
use std::io::{Write, Read, Cursor};

pub struct BcsTextStream {
    buffer: Mutex<Cursor<Vec<u8>>>,
}

impl BcsTextStream {
    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            buffer: Mutex::new(Cursor::new(Vec::new())),
        })
    }

    pub fn write_string(&self, text: &str) {
        let mut buf = self.buffer.lock().unwrap();
        let _ = buf.write_all(text.as_bytes());
    }

    pub fn read_all(&self) -> String {
        let mut buf = self.buffer.lock().unwrap();
        let mut result = String::new();
        let pos = buf.position();
        buf.set_position(0);
        let _ = buf.read_to_string(&mut result);
        buf.set_position(pos);
        result
    }

    pub fn flush(&self) {
        let mut buf = self.buffer.lock().unwrap();
        let _ = buf.flush();
    }
}
